src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -32,15 +38,91 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory, IDispo
-        public ProjectCodeModelFactory(VisualStudioWorkspace visualStudioWorkspace, [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, IThreadingContext threadingContext)
+        public ProjectCodeModelFactory(
+            VisualStudioWorkspace visualStudioWorkspace,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IThreadingContext threadingContext,
+            IForegroundNotificationService notificationService,
+            IAsynchronousOperationListenerProvider listenerProvider)
         {
             _visualStudioWorkspace = visualStudioWorkspace;
             _serviceProvider = serviceProvider;
             _threadingContext = threadingContext;
             _deferredCleanupTasks = new JoinableTaskCollection(threadingContext.JoinableTaskContext);
             _deferredCleanupTasks.DisplayName = nameof(ProjectCodeModelFactory) + "." + nameof(_deferredCleanupTasks);
+
+            _notificationService = notificationService;
+            _listener = listenerProvider.GetListener(FeatureAttribute.CodeModel);
+
+            // Queue up notifications we hear about docs changing.  that way we don't have to fire events multiple times
+            // for the same documents.  Once enough time has passed, take the documents that were changed and run
+            // through them, firing their latest events.
+            _documentsToFireEventsFor = new AsyncBatchingWorkQueue<DocumentId>(
+                TimeSpan.FromMilliseconds(visualStudioWorkspace.Options.GetOption(InternalSolutionCrawlerOptions.AllFilesWorkerBackOffTimeSpanInMS)),
+                ProcessNextDocumentBatchAsync,
+                // We only care about unique doc-ids, so pass in this comparer to collapse streams of changes for a
+                // single document down to one notification.
+                EqualityComparer<DocumentId>.Default,
+                _listener,
+                _cancellationTokenSource.Token);
+
+            _visualStudioWorkspace.WorkspaceChanged += OnWorkspaceChanged;
+        }
+
+        private System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
+            ImmutableArray<DocumentId> documentIds, CancellationToken cancellationToken)
+        {
+            foreach (var documentId in documentIds)
+            {
+                // Now, enqueue foreground work to actually process these documents in a serialized and incremental
+                // fashion.  FireEventsForDocument will actually limit how much time it spends firing events so that it
+                // doesn't saturate  the UI thread.
+                _notificationService.RegisterNotification(
+                    () => FireEventsForDocument(documentId),
+                    _listener.BeginAsyncOperation("CodeModelEvent"),
+                    cancellationToken);
+            }
+
+            return System.Threading.Tasks.Task.CompletedTask;
```

Comment:
>System.Threading.Tasks [](start = 19, length = 22)

`using System.Threading.Tasks`

Reply:
can't.  ambiguous with Task from Microsoft.VisualStudio :-/  annoying for sure.  If we had more cases in the file, i'd likely alias.  BUt there so few usages, i don't mind just being explicit.

Reply:
Then I'd prefer to qualify VS types and not import the VS namespace that contains Task over qualifying Task.